// ompimon-protocol/test/client/testCounter.js
var
    BufferBuilder = require("buffer-builder"),
    _ = require("underscore"),
    binary = require("binary"),
    Int64 = require('node-int64'),

    Client = require("ompimon-client/client"),

    Protocol = require("../.."),
    auth = require("ompimon-auth").auth,
    Parser = require("../../parser"),
    stub = require("../../test-client-stub"),

    action = require("../../client_actions/Counter"),

    clusterAction = require("../../cluster_actions/Data"),

    clusterStub = require("../../test-stub")


    ;
auth.type = "test";

var Action = new action;

var client = new Client('1', new Buffer(256));

module.exports = {

    testParse: function(test) {
        var data = stub.buildCounter(stub.counterData);
        var parser = new Parser(data);

        var result = Action.parse(client, parser);

        test.deepEqual(result, stub.counterData);
        test.done();
    },

    testParseOptional: function(test) {
        var data2 = _.clone(stub.counterData);
        data2.counterId = null;

        var data = stub.buildCounter(data2);
        var parser = new Parser(data);

        var result = Action.parse(client, parser);
        test.deepEqual(result, data2);
        test.done();
    },

    testProcess: function (test) {
        Action.process(client, {
            appId: 0,
            counterId: null
        });

        Action.once('send', function(client, data) {
            console.log(data);
            test.notEqual(data, null);
            test.done();
        });
    },

    testProcess2: function (test) {
        Action.process(client, {
            appId: 0,
            counterId: 0
        });

        Action.once('send', function(client, data) {
            console.log(data);
            test.notEqual(data, null);
            test.done();
        });
    },
    testProcessAsync: function (test) {
        Action.process(client, {
            appId: 0,
            counterId: null
        });

        Action.once('send', function(client, data) {
            console.log(data);
            test.notEqual(data, null);

            var parser = new Parser(clusterStub.buildSend(clusterStub.sendData).get());

            var result = clusterAction.parse(client, parser);


        });
    }
};
