// ompimon-protocol/actions/init.js
/**
 * @module ompimon
 * @submodule ompimon-protocol
 */

var
    util = require("util"),
    _ = require("underscore"),
    binary = require("binary"),
    BufferBuilder = require('buffer-builder'),

    auth = require('ompimon-auth').auth,

    action = require("./action"),
    Parser = require("../parser"),
    response = require("../response"),
    globals = require("../globals")
;

module.exports = Init;


/**
 * @class Init
 * @namespace Ompimon.Protocol.Action
 * @extends Ompimon.Protocol.Action.Action
 * @constructor
 */
function Init() {
    action.call(this);
    /**
     * @property needAuthentication
     * @type {boolean}
     * @default false
     */
    this.needAuthentication = false;
}

util.inherits(Init, action);

_.extend(Init.prototype, {
    /**
     * Parse data send by cluster.
     * Will return something like:

     {
         protocolVersion: 0x01,
         username: "test",
         password: "test",
         app: "Meine Tolle Applikation",
         processes: 123782,
         nodes: 1010000,
         nodeId: 0,
         ranks: [
             123, 495
         ],
         counters: [
             "broadcast",
             "barrier"
         ],
         sends: [
             "ibsend",
             "bsend",
             "irsend"
         ],
         features: {
             restart: true,
             abort: true
         }
     }

     * @method parse
     * @param client {Ompimon.Cluster.Client}
     * @param parser {Ompimon.Protocol.Parser}
     * @return {Object}
     */
    parse: function(client, parser) {
        var result = {
            protocolVersion: parser.readUInt32(),
            username:        parser.readString(parser.readUInt8()),
            password:        parser.readString(parser.readUInt8()),
            app:             parser.readString(parser.readUInt16()),
            processes:       parser.readUInt32(),
            nodes:           parser.readUInt32(),
            nodeId:          parser.readUInt32(),
            ranks:           this._parseRanks(parser)
        };
        if (result.nodeId == 0) {
            _.extend(result, {
                counters:   this._parseCounters(parser),
                sends:      this._parseSends(parser),
                features:   this._parseFeatures(parser)
            });

        }

        return result;
    },
    _parseRanks: function(parser) {
        var ranks = [];
        var count = parser.readUInt32();
        for (var i = 0; i < count; i++) {
            ranks.push(parser.readUInt32());
        }

        return ranks;
    },
    _parseCounters: function(parser) {
        var counters = [];
        var len = parser.readUInt32();
        for (var i = 0; i < len; i++) {
            counters.push(parser.readString(parser.readUInt8()));
        }

        return counters;
    },
    _parseSends: function(parser) {
        var sends = [];
        var len = parser.readUInt32();
        for (var i = 0; i < len; i++) {
            sends.push(parser.readString(parser.readUInt8()));
        }

        return sends;
    },
    _parseFeatures: function(parser) {
        var features = {};
        _.each(globals.features, function(feature) {
            features[feature] = false;
        });

        var count = parser.readUInt8();
        for (var i = 0; i < count; i++) {
            var feature = globals.features[parser.readUInt8()];
            features[feature] = true;
        }

        return features;
    },
    process: function(client, data, callback) {
        this._authenticate(client, data.username, data.password, function(result) {
            if (result.status != response.OK) {
                callback(result);

                return;
            }

            client.counterFunctions = data.counters;
            client.sendFunctions = data.sends;
            client.ranks = data.ranks;

            //@TODO mega geile action (lukas)
            callback(new response.Response(response.OK));
        });
    },
    _authenticate: function(client, username, password, callback) {

        auth.authenticate("cluster", username, password, function() {
            client.authenticated = true;
            callback(new response.Response(response.OK));
        }, function() {
            client.authenticated = false;
            callback(new response.Response(response.NOT_AUTHENTICATED));
        });
    }

});
