// ompimon-protocol/parser.js
var Int64 = require('node-int64')
;

module.exports = Parser;

function Parser(buffer) {
    this.offset = 0;
    this.buffer = buffer;
}

Parser.prototype = {
    readUInt8: function() {
        return this.buffer.readUInt8(this.offset++);
    },
    readUInt16: function() {
        var result = this.buffer.readUInt16BE(this.offset);
        this.offset += 2;

        return result;
    },
    readUInt32: function() {
        var result = this.buffer.readUInt32BE(this.offset);
        this.offset += 4;

        return result;
    },
    readUInt64: function() {
        return new Int64(this.readBuffer(8));
    },
    readString: function(length) {
        return this.readBuffer(length).toString('utf8');
    },
    readBuffer: function(length) {
        var buffer = this.buffer.slice(this.offset, this.offset + length);
        this.offset += length;

        return buffer;
    },
    debug: function() {
        console.log("Buffer");
        console.log(this.buffer.slice(this.offset));
        console.log("----");
    }

};
