// ompimon-protocol/test-stub.js
var
    BufferBuilder = require("buffer-builder"),
    _ = require("underscore"),
    binary = require("binary"),
    Int64 = require('node-int64'),
    globals = require("./globals")

;

module.exports = {
    initData: {
        protocolVersion: 0x01,
        username: "test",
        password: "install",
        app: "Meine Tolle Applikation",
        processes: 123782,
        nodes: 1010000,
        nodeId: 0,
        ranks: [
            123, 495
        ],
        counters: [
            "broadcast",
            "barrier"
        ],
        sends: [
            "ibsend",
            "bsend",
            "irsend"
        ],
        features: {
            restart: true,
            abort: true
        }
    },

    buildInit: function (data) {
        var buffer = new BufferBuilder();

        buffer
            .appendUInt32BE(data.protocolVersion)
            // username
            .appendUInt8(data.username.length)
            .appendString(data.username)
            // password
            .appendUInt8(data.password.length)
            .appendString(data.password)
            // app name
            .appendUInt16BE(data.app.length)
            .appendString(data.app)
            // processes
            .appendUInt32BE(data.processes)
            // nodes
            .appendUInt32BE(data.nodes)
            // nodeId
            .appendUInt32BE(data.nodeId)
            // rank count
            .appendUInt32BE(data.ranks.length)
        ;
        // ranks
        data.ranks.forEach(function (rank) {
            buffer.appendUInt32BE(rank);
        });

        if (data.nodeId != 0) {
            return buffer;
        }

        // counter functions
        buffer.appendUInt32BE(data.counters.length);
        data.counters.forEach(function (counter) {
            buffer.appendUInt8(counter.length);
            buffer.appendString(counter);
        });
        //send/receive functions
        buffer.appendUInt32BE(data.sends.length);
        data.sends.forEach(function (send) {
            buffer.appendUInt8(send.length);
            buffer.appendString(send);
        });

        //features
        var features = [];
        _.each(data.features, function (value, feature) {
            var id = _.invert(globals.features)[feature];
            if (value) {
                features.push(id);
            }
        });
        buffer.appendUInt8(features.length);
        features.forEach(function (feature) {
            buffer.appendUInt8(feature);
        });


        return buffer;
    },


    sendData: [
        {
            rank: 13,
            counters: {
                broadcast: 123123,
                barrier: 4234234
            },
            data: [
                {
                    rank: 2,
                    counter: new Int64(12344),
                    size: new Int64(237373)
                },
                {
                    rank: 3,
                    counter: new Int64(123123),
                    size: new Int64(2020203)
                }
            ]
        },
        {
            rank: 24,
            counters: {
                broadcast: 595959,
                barrier: 202340234
            },
            data: [
                {
                    rank: 1,
                    counter: new Int64(123123855),
                    size: new Int64(2034)
                }
            ]
        }
    ],

    buildSend: function (data) {
        var buffer = new BufferBuilder();
        // processes
        data.forEach(function(dataEntry) {
            // counters
            _.each(dataEntry.counters, function(counter) {
                buffer.appendUInt32BE(counter);
            });

            // data
            buffer.appendUInt32BE(dataEntry.data.length);
            dataEntry.data.forEach(function(dataSet) {
                // rank id
                buffer.appendUInt32BE(dataSet.rank);
                // counter
                buffer.appendBuffer(dataSet.counter.buffer);
                // size
                buffer.appendBuffer(dataSet.size.buffer);
            });
        });

        return buffer;
    },

    requestData: [
        {
            rank: 13,
            functions: {
                ibsend: [
                    {
                        rank: 2,
                        counter: new Int64(123),
                        size: new Int64(1231234)
                    },
                    {
                        rank: 3,
                        counter: new Int64(47474),
                        size: new Int64(2349)
                    }
                ],
                bsend: [
                    {
                        rank: 4,
                        counter: new Int64(2349),
                        size: new Int64(239492)
                    }
                ]
            }
        },
        {
            rank: 24,
            functions: {
                ibsend: [
                    {
                        rank: 1,
                        counter: new Int64(123123),
                        size: new Int64(349234)
                    }
                ],
                bsend: [
                ]
            }
        }
    ],

    buildDataRequest: function (data) {
        var buffer = new BufferBuilder();
        data.forEach(function(rank) {
            _.each(rank.functions, function(func) {
                // array length
                buffer.appendUInt32BE(func.length);
                func.forEach(function(funcData) {
                    //rank
                    buffer.appendUInt32BE(funcData.rank);
                    // counter
                    buffer.appendBuffer(funcData.counter.buffer);
                    // size
                    buffer.appendBuffer(funcData.size.buffer);
                });
            });
        });

        return buffer;
    }
};
