// ompimon-protocol/test-client-stub.js
var
    BufferBuilder = require("buffer-builder"),
    _ = require("underscore")
;

module.exports = {
    loginData: {
        protocolVersion: 0x01,
        username: "test",
        password: "install"
    },

    buildLogin: function (data) {
        var buffer = new BufferBuilder();
        buffer.appendUInt8(data.protocolVersion);
        // username
        buffer.appendUInt8(data.username.length);
        buffer.appendString(data.username);
        // password
        buffer.appendUInt8(data.password.length);
        buffer.appendString(data.password);

        return buffer.get();
    },

    restartData: {
        appId: 123
    },

    buildRestart: function (data) {
        var buffer = new BufferBuilder();
        buffer.appendUInt32BE(data.appId);

        return buffer.get();
    },

    abortData: {
        appId: 123
    },

    buildAbort: function (data) {
        var buffer = new BufferBuilder(data.appId);
        buffer.appendUInt32BE(data.appId);

        return buffer.get();
    }
};
