// ompimon-protocol/test-client-stub.js
var
    BufferBuilder = require("buffer-builder"),
    _ = require("underscore")
;

module.exports = {
    loginData: {
        protocolVersion: 0x01,
        username: "test",
        password: "install"
    },

    buildLogin: function (data) {
        var buffer = new BufferBuilder();
        buffer.appendUInt8(data.protocolVersion);
        // username
        buffer.appendUInt8(data.username.length);
        buffer.appendString(data.username);
        // password
        buffer.appendUInt8(data.password.length);
        buffer.appendString(data.password);

        return buffer.get();
    },

    applicationData: {},

    buildApplication: function (data) {
        var buffer = new BufferBuilder();

        return buffer.get();
    },

    appDetailData: {
        appId: 123
    },

    buildAppDetail: function (data) {
        var buffer = new BufferBuilder();
        buffer.appendUInt32BE(data.appId);

        return buffer.get();
    },

    counterData: {
        appId: 123,
        counterId: 2 // optional
    },

    buildCounter: function (data) {
        var buffer = new BufferBuilder();
        buffer.appendUInt32BE(data.appId);
        if (data.counterId) {
            buffer.appendUInt8(1);
            buffer.appendUInt32BE(data.counterId);
        } else {
            buffer.appendUInt8(0);
        }
        return buffer.get();
    },

    counterDetailData: {
        appId: 123,
        processId: 456,
        counterId: 123939
    },

    buildCounterDetail: function (data) {
        var buffer = new BufferBuilder();
        buffer.appendUInt32BE(data.appId);
        buffer.appendUInt32BE(data.processId);
        if (data.counterId) {
            buffer.appendUInt8(1);
            buffer.appendUInt32BE(data.counterId);
        } else {
            buffer.appendUInt8(0);
        }

        return buffer.get();
    },

    sendData: {
        appId: 123,
        sendId: 4949 // optinal
    },

    buildSend: function (data) {
        var buffer = new BufferBuilder();
        buffer
            .appendUInt32BE(data.appId)
        ;
        if (data.sendId) {
            buffer.appendUInt8(1);
            buffer.appendUInt32BE(data.sendId);
        } else {
            buffer.appendUInt8(0);
        }

        return buffer.get();
    },

    sendDetailData: {
        appId: 123,
        processId: 4949,
        sendId: 456 // optional
    },

    buildSendDetail: function (data) {
        var buffer = new BufferBuilder();
        buffer
            .appendUInt32BE(data.appId)
            .appendUInt32BE(data.processId)
        ;
        if (data.sendId) {
            buffer.appendUInt8(1);
            buffer.appendUInt32BE(data.sendId);
        } else {
            buffer.appendUInt8(0);
        }

        return buffer.get();
    }

};
