// ompimon-cluster/BufferStack.js
/**
 * @module ompimon
 * @submodule ompimon-cluster
 */
var
    util = require("util"),
    eventEmitter = require('events').EventEmitter,
    Buffer = require("buffer").Buffer,
    _ = require("underscore")
;

module.exports = BufferStack;

/**
 * Holds buffers received by a client
 * @class BufferStack
 * @namespace Ompimon.Cluster
 * @constructor
 */
function BufferStack() {
    eventEmitter.call(this);
}

util.inherits(BufferStack, eventEmitter);

_.extend(BufferStack.prototype, {
    /**
     * Holds a buf
     * @private
     * @property stack
     * @type Buffer
     */
    stack: null,
    /**
     * Push a new buffer to the stack
     * fires an pushed event
     * @param buffer {Buffer}
     * @method push
     *
     */
    push: function(buffer) {
        if (null === this.stack) {
            this.stack = buffer;
        } else {
            this.stack = Buffer.concat([this.stack, buffer]);
        }
        this.emit('pushed', this);
    },

    getBuffer: function() {
        var stack = this.stack;
        this.stack = null;
        return stack;
    },
    /**
     * returns true if stack is empty
     * @method empty
     * @return boolean
     */
    empty: function() {
        return this.stack.length == 0;
    },
    prepend: function (buffer) {
        if (null == this.stack) {
            this.stack = buffer;
        } else {
            this.stack.slice(0).unshift(buffer);
        }
    },
    /**
     * Will be fired if a new buffer is pushed to the stack
     *
     * @event pushed
     */
    pushed: function() {}
});
