// ompimon-cluster/worker.js
/**
 * @module ompimon
 * @submodule ompimon-cluster
 */

var eventEmitter = require('events').EventEmitter,
    net = require('net'),
    client = require('./client'),
    Protocol = require('ompimon-protocol'),
    crypto = require("crypto"),
    BufferBuilder = require('buffer-builder'),

    util = require("util"),
    _ = require("underscore")
;

var protocol = new Protocol();

protocol.on('send', function(client, data) {
   client.send(data);
});

module.exports = Worker;

/**
 * @class Worker
 * @namespace Ompimon.Cluster
 *
 * @param port {Integer} Port this worker listens to
 * @param [start=true] {Boolean} Whether to start the worker immediately
 * @constructor
 */
function Worker (port, start) {
    start = typeof start !== 'undefined' ? start : true;
    eventEmitter.call(this);
    /**
     * Port this worker listens to
     * @property port
     * @type {Integer}
     */
    this.port = port;
    if (start) {
        this.start();
    }
}

util.inherits(Worker, eventEmitter);

_.extend(Worker.prototype, {
    /**
     * Start listening to incomming connections
     *
     * @method start
     */
    start: function() {
        var self = this;
        var server = net.createServer(function (c) { //'connection' listener
            var clientInstance = new client(c);

            self.connect(clientInstance);
            c.on('end', function () {
                console.log('client '+clientInstance.id+' disconnected');
                clientInstance.quit();
                clientInstance = null;
            });

        });


        server.listen(this.port, function () { //'listening' listener
            console.log("Server started on port " + self.port);
        });
    },

    /**
     * Handle connection of new clients
     *
     * @method connect
     * @param client {Ompimon.Cluster.Client}
     */
    connect: function(client) {
        var self = this;

        console.log("client "+ client.id+" connected");

        client.socket.on('data', function(data) {
            self.process(client, data);
        });

        client.buffers.on('pushed', function(buffers) {
            protocol.parseCluster(client);
        });
    },

    /**
     * Process received messages
     *
     * @method process
     * @param client {Ompimon.Cluster.process}
     * @param data {Buffer}
     * @param [callback] {Function}
     */
    process: function(client, data, callback) {
        var string = data.toString('utf-8').trim().replace(/(\r\n|\n|\r)/gm,"");
        console.log("received from " + client.id+":");
        console.log(data);
        //console.log(string);

        if (callback) {
            client.once('parsed', callback);
        }

        client.buffers.push(data);


        /*
        try {
            Protocol.parse(client, data, function(buf) {
                if (null !== buf) {
                    client.socket.write(buf);
                }
                if (callback) { callback(); }

            });
        } catch (e) {
            console.log(e);
            var result = new BufferBuilder();
            result.appendUInt8(0x02); // invalid input

            client.socket.write(result.get());
            if (callback) { callback() }
            throw e;
        }

        setTimeout(function() {
        }, 500);*/
    }
});
