// ompimon-storage/main.js
var _ = require("underscore")._,
    redis = require("redis"),
    rclient = redis.createClient();

var init = false;

exports.init = function(){
  init = true;
};

/**
 *
 * @param object data
 * @returns int | boolean
 */
exports.createApplication = function(data){
    if(data.protocolVersion != 1){
        return false;
    }
    var applen = rclient.LLEN("apps");
    var found = false;

    for(var i = 0; i < applen; i++){
        var tempAppName = rclient.LINDEX("apps", i);
        if(tempAppName == data.app){
            return i;
        }
    }

    if(!found){
        rclient.RPUSH("apps", data.app);
        rclient.HMSET("app_"+applen,{
            processes: data.processes,
            nodes: data.nodes,
            restart: data.features.restart,
            abort: data.features.abort

        });
        data.ranks.forEach(function(value) {
            rclient.SADD("app_"+applen+"_ranks", value);
        });

        data.counters.forEach(function(value) {
            rclient.SADD("app_"+applen+"_counters", value);
        });

        data.sends.forEach(function(value) {
            rclient.SADD("app_"+applen+"_sends", value);
        });

        data.features.forEach(function(value) {
            rclient.SADD("app_"+applen+"_features", value);
        });

        return applen;
    }
};


/*
var data = {
    protocolVersion: 0x01,
    username: "test",
    password: "test",
    app: "Meine Tolle Applikation",
    processes: new Int64(123782),
    nodes: new Int64(1010000),
    nodeId: 0,
    ranks: [
        123, 495
    ],
    counters: [
        "broadcast",
        "barrier"
    ],
    sends: [
        "ibsend",
        "bsend",
        "irsend"
    ],
    features: {
        restart: true,
        abort: true
    }
};*/

exports.receive = function(data){

};

/*var data = {
    "lists": [
        {
            "sender": 123,
            "receivers": [
                {
                    "id": 456,
                    "messages": '0000000000000010',
                    "data": 'ffffffffffffffff'
                },
                {
                    "id": 968,
                    "messages": '00000000000000bf',
                    "data": '00000000000000af'
                },
                {
                    "id": 2004,
                    "messages": '0000000000002020',
                    "data": '0000000000282349'
                }
            ]
        },
        {
            "sender": 456,
            "receivers": [
                {
                    "id": 123,
                    "messages": '0000000000003847',
                    "data": '0000000000002929'
                }
            ]
        }
    ]
};*/
