// ompimon-monitor/monitor.js
var
    _ = require("underscore"),
    util = require("util"),
    color = require("bash-color")
;

module.exports = Monitor;

function Monitor(type, name)
{
    this.init.apply(this, arguments);
}

_.extend(Monitor.prototype, {
    init: function (type, name) {
        if (!config.monitor) {
            return;
        }
        this.type = type;
        this.name = name;
        this.log("started");
        this.addComponent(type, name);
    },

    message: function(from, to, message, payload) {
        if (!config.monitor) {
            return;
        }
        this.log("from " + from + " to "+to+":" + message);
    },

    addComponent: function(type, name) {
        if (!config.monitor) {
            return;
        }
        this.log("added " + type+":"+name);
    },

    removeComponent: function(type, name) {
        if (!config.monitor) {
            return;
        }

        this.log("removed "+ type+":"+name);
    },

    log: function (message) {
        console.log(color.purple("[" + this.type+":"+ this.name+ "]") + ": " + message);
    }
});
