// ompimon-protocol/client_actions/Abort.js
/**
 * @module ompimon
 * @submodule ompimon-protocol
 */

var
    util = require("util"),
    _ = require("underscore"),

    auth = require('ompimon-auth').auth,

    Storage = require("ompimon-storage"),
    Redis = require("redis"),

    parser = require("../parser"),
    response = require("../response"),

    action = require("./Action")
;


module.exports = Abort;

/**
 * @class Abort
 * @namespace Ompimon.Protocol.ClientAction
 * @extends Ompimon.Protocol.ClientAction.Action
 * @constructor
 */
function Abort() {
    action.call(this);
}

util.inherits(Abort, action);

_.extend(Abort.prototype, {
    /**
     * Parse a message
     * Will return something like:

     {
        protocolVersion: 0x01,
        username: "test",
        password: "install"
     }

     * @method parse
     * @param client {Websocket}
     * @param parser {Ompimon.Protocol.Parser}
     */
    parse: function (client, parser) {
        return {
            appId: parser.readUInt32()
        }
    },
    /**
     * processes parsed data
     * @method process
     * @param client {Websocket}
     * @param data {Object}
     */
    process: function (client, data) {
        var self = this;
        var message = {action: 'abort'};

        var redisClient = Redis.createClient();

        var listenChannel = "client:app:"+data.appId+":abort";

        redisClient.on('subscribe', function (channel, count) {
            console.log("subscribed to " + channel, count);
        });

        redisClient.subscribe(listenChannel);


        redisClient.on('message', function(channel, data) {
            if (channel != listenChannel) {
                return false;
            }
            data = JSON.parse(data);
            if (data.status == 'ok') {
                redisClient.quit();
                self.emit('send', client, new response.Response(response.OK));
            }
        });

        var client2 = Redis.createClient();
        client2.publish("cluster:app:" + data.appId + ":abort", JSON.stringify(message));
        client2.quit();
        console.log("pub");
    }
});
