// ompimon-protocol/client_actions/Applications.js
/**
 * @module ompimon
 * @submodule ompimon-protocol
 */

var
    util = require("util"),
    _ = require("underscore"),
    auth = require('ompimon-auth').auth,
    BufferBuilder = require("buffer-builder"),
    Storage = require("ompimon-storage"),

    parser = require("../parser"),
    response = require("../response"),


    action = require("./Action")
;

module.exports = Applications;

/**
 * Request a list of all running applications
 *
 * @class Applications
 * @namespace Ompimon.Protocol.ClientAction
 * @extends Ompimon.Protocol.ClientAction.Action
 * @constructor
 * @author Florian Pfitzer<pfitzer@w3p.cc>
 */
function Applications() {
    action.call(this);
}

util.inherits(Applications, action);

_.extend(Applications.prototype, {
    /**
     * Parse a message
     * Will return something like:

     {
     }

     * @method parse
     * @param client {Ompimon.Client.Client}
     * @param parser {Ompimon.Protocol.Parser}
     */
    parse: function (client, parser) {
        return {}; // no payload
    },
    /**
     * processes parsed data
     * @method process
     * @param client {Ompimon.Client.Client}
     * @param data {Object}
     */
    process: function (client, data) {
        this._sendApplications(client);

        // listen for application list changes
        client.on('app_list', function(message) {
            this._sendApplications(client);
        }.bind(this));
    },

    /**
     * Send application list to a client
     *
     * @method sendApplications
     * @param client
     * @private
     */
    _sendApplications: function(client) {
        monitor.message('me', 'db', 0x02);

        Storage.getApplications(function(err, data) {
            if (err) {
                console.log(err);
            } else {
                monitor.message('db', 'me', 0x02, data);
                this.emit('send', client, this.buildData(data));
            }
        }.bind(this));
    },

    /**
     * Will parse an object to binary

     [
         {
             id: 123,
             name: "Foo"
         },
         {
             id: 349,
             name: "Bar"
         }
     ]

     * @method buildData
     * @param data {Object}
     * @return Buffer
     */
    buildData: function (data) {
        var buffer = new BufferBuilder();
        buffer.appendUInt8(data.length); // numberOfApps
        _.each(data, function (app, key) {
            buffer.appendUInt32BE(app.id); // appId
            buffer.appendUInt32BE(app.name.length); // length
            buffer.appendString(app.name);
        });

        return buffer.get();
    }

});
