// ompimon-protocol/client_actions/Applications.js
/**
 * @module ompimon
 * @submodule ompimon-protocol
 */

var
    util = require("util"),
    _ = require("underscore"),
    auth = require('ompimon-auth').auth,
    BufferBuilder = require("buffer-builder"),

    parser = require("../parser"),
    response = require("../response"),


    action = require("./Action")
;

module.exports = Applications;

/**
 * @class Applications
 * @namespace Ompimon.Protocol.ClientAction
 * @extends Ompimon.Protocol.ClientAction.Action
 * @constructor
 */
function Applications() {
    action.call(this);
}

util.inherits(Applications, action);

_.extend(Applications.prototype, {
    /**
     * Parse a message
     * Will return something like:

     {
     }

     * @method parse
     * @param client {Websocket}
     * @param parser {Ompimon.Protocol.Parser}
     */
    parse: function (client, parser) {
        return {}; // no payload
    },
    /**
     * processes parsed data
     * @param client {Websocket}
     * @param data {Object}
     */
    process: function (client, data) {
        // TODO implement @Flo
    },

    /**
     * Will parse an object to binary

     [
         {
             id: 123,
             name: "Foo"
         },
         {
             id: 349,
             name: "Bar"
         }
     ]

     * @method buildData
     * @param data {Object}
     * @return Buffer
     */
    buildData: function (data) {
        var buffer = new BufferBuilder();
        buffer.appendUInt8(data.length); // numberOfApps
        data.forEach(function (app) {
            buffer.appendUInt32BE(app.id); // appId
            buffer.appendUInt32BE(app.name.length); // length
            buffer.appendString(app.name);
        });

        return buffer.get();
    }

});
