// ompimon-protocol/client_actions/CounterDetail.js
/**
 * @module ompimon
 * @submodule ompimon-protocol
 */

var
    util = require("util"),
    _ = require("underscore"),
    auth = require('ompimon-auth').auth,


    parser = require("../parser"),
    response = require("../response"),


    action = require("./Action")
    ;

module.exports = CounterDetail;

/**
 * @class CounterDetail
 * @namespace Ompimon.Protocol.ClientAction
 * @extends Ompimon.Protocol.ClientAction.Action
 * @constructor
 */
function CounterDetail() {
    action.call(this);
}

util.inherits(CounterDetail, action);

_.extend(CounterDetail.prototype, {
    /**
     * Parse a message
     * Will return something like:

     {
        appId: 123,
        processId: 456,
        counterId: 123939
     }

     * @method parse
     * @param client {Websocket}
     * @param parser {Ompimon.Protocol.Parser}
     */
    parse: function (client, parser) {

        return {
            appId: parser.readUInt32(),
            processId: parser.readUInt32(),
            counterId: (parser.readUInt8()) ? parser.readUInt32() : null
        }
    },


    /**
     * processes parsed data
     * @param client {Websocket}
     * @param data {Object}
     */
    process: function (client, data) {
        // TODO implement @Flo
    },

    /**
     * Will parse an object to binary

     [
         {
             time: 1234737127, // unix timestamp (seconds)
             counter: 1239239
         },
         {
             time: 28423742234,
             counter: 1237123
         }
     ]

     * @method buildData
     * @param data {Object}
     * @return Buffer
     */
    buildData: function (data) {
        var buffer = new BufferBuilder();

        buffer.appendUInt16BE(data.length); // length

        data.forEach(function (process) {
            buffer.appendUInt32BE(process.time);
            buffer.appendUInt32BE(process.counter);
        });

        return buffer.get();
    }

});
