// ompimon-protocol/client_actions/Login.js
/**
 * @module ompimon
 * @submodule ompimon-protocol
 */

var
    util = require("util"),
    _ = require("underscore"),
    BufferBuilder = require("buffer-builder"),
    auth = require('ompimon-auth').auth,


    parser = require("../parser"),
    response = require("../response"),


    action = require("./Action")
;


module.exports = Login;

/**
 * Authenticate against the backend. The most actions are only available for
 * authenticated users.
 *
 * @class Login
 * @namespace Ompimon.Protocol.ClientAction
 * @extends Ompimon.Protocol.ClientAction.Action
 * @constructor
 * @author Florian Pfitzer<pfitzer@w3p.cc>
 */
function Login() {
    action.call(this);
}

util.inherits(Login, action);

_.extend(Login.prototype, {
    /**
     * Whether this client needs authentication
     * @property needAuthentication
     * @type boolean
     * @default false
     */
    needAuthentication: false,
    /**
     * Parse a message
     * Will return something like:

     {
        protocolVersion: 0x01,
        username: "test",
        password: "install"
     }

     * @method parse
     * @param client {Ompimon.Client.Client}
     * @param parser {Ompimon.Protocol.Parser}
     */
    parse: function (client, parser) {
        return {
            protocolVersion: parser.readUInt8(),
            username: parser.readString(parser.readUInt8()),
            password: parser.readString(parser.readUInt8())
        }
    },
    /**
     * processes parsed data
     * @method process
     * @param client {Ompimon.Client.Client}
     * @param data {Object}
     */
    process: function (client, data) {
        var self = this;
        this._authenticate(client, data.username, data.password, function(result) {
            if (result.status != response.OK) {
                self.emit('send', client, result.status);

                return;
            }

            var res = new BufferBuilder();
            res.appendUInt8(0x00);

            self.emit('send', client, res.get());
        });
    },

    /**
     * Authenticate a user against the backend
     * @method authenticate
     * @param client {Ompimon.Client.Client}
     * @param username {String}
     * @param password {String}
     * @param callback {Function}
     * @private
     * @async
     */
    _authenticate: function(client, username, password, callback) {
        auth.authenticate("android", username, password, function() {
            client.authenticated = true;
            callback(new response.Response(response.OK));
        }, function() {
            client.authenticated = false;
            callback(new response.Response(response.NOT_AUTHENTICATED));
        });
    }
});
