// ompimon-protocol/client_actions/Send.js
/**
 * @module ompimon
 * @submodule ompimon-protocol
 */

var
    util = require("util"),
    _ = require("underscore"),
    auth = require('ompimon-auth').auth,


    parser = require("../parser"),
    response = require("../response"),


    action = require("./Action")
    ;

module.exports = Send;

/**
 * @class Send
 * @namespace Ompimon.Protocol.ClientAction
 * @extends Ompimon.Protocol.ClientAction.Action
 * @constructor
 */
function Send() {
    action.call(this);
}

util.inherits(Send, action);

_.extend(Send.prototype, {
    /**
     * Parse a message
     * Will return something like:

     {
        appId: 123,
        sendId: 4949 // optinal
     }

     * @method parse
     * @param client {Websocket}
     * @param parser {Ompimon.Protocol.Parser}
     */
    parse: function (client, parser) {
        //TODO implement @Patricio
    },


    /**
     * processes parsed data
     * @param client {Websocket}
     * @param data {Object}
     */
    process: function (client, data) {
        // TODO implement @Flo
    },

    /**
     * Will parse an object to binary

     [
         {
             senderId: 4848,
             receiverId: 4949,
             bytes: Int64(202)
         },
         {
             senderId: 49494,
             receiverId: 494949,
             bytes: Int64(4949)
         }
     ]

     * @method buildData
     * @param data {Object}
     * @return Buffer
     */
    buildData: function (data) {
        var buffer = new BufferBuilder();

        buffer.appendUInt16BE(data.length); // length

        data.forEach(function (process) {
            buffer
                .appendUInt32BE(process.senderId)
                .appendUInt32BE(process.receiverId)
                .appendBuffer(process.bytes.buffer)
            ;
        });

        return buffer.get();
    }

});
