// ompimon-protocol/client_actions/SendDetail.js
/**
 * @module ompimon
 * @submodule ompimon-protocol
 */

var
    util = require("util"),
    _ = require("underscore"),
    auth = require('ompimon-auth').auth,


    parser = require("../parser"),
    response = require("../response"),


    action = require("./Action")
    ;

module.exports = CounterDetail;

/**
 * @class SendDetail
 * @namespace Ompimon.Protocol.ClientAction
 * @extends Ompimon.Protocol.ClientAction.Action
 * @constructor
 */
function CounterDetail() {
    action.call(this);
}

util.inherits(CounterDetail, action);

_.extend(CounterDetail.prototype, {
    /**
     * Parse a message
     * Will return something like:

     {
        appId: 123,
        processId: 4949,
        sendId: 456, // optional
     }

     * @method parse
     * @param client {Websocket}
     * @param parser {Ompimon.Protocol.Parser}
     */
    parse: function (client, parser) {

        return {
            appId: parser.readUInt32(),
            processId: parser.readUInt32(),
            sendId: (parser.readUInt8()) ? parser.readUInt32() : null
        }

    },


    /**
     * processes parsed data
     * @param client {Websocket}
     * @param data {Object}
     */
    process: function (client, data) {
        // TODO implement @Flo
    },

    /**
     * Will parse an object to binary

     [
         {
             time: 9234912491, // timestamp in unix seconds,
             processId: 184824,
             bytes: Int64(12391239)
         },
         {
             time: 12301230123,
             processId: 49494,
             bytes: Int64(2340234)
         }
     ]

     * @method buildData
     * @param data {Object}
     * @return Buffer
     */
    buildData: function (data) {
        var buffer = new BufferBuilder();

        buffer.appendUInt16BE(data.length); // length

        data.forEach(function (process) {
            buffer
                .appendUInt32BE(process.time)
                .appendUInt32BE(process.processId)
                .appendBuffer(process.bytes.buffer)
            ;
        });

        return buffer.get();
    }
});
