// ompimon-storage/test/01_clusterInitTest.js
/**
 * @author Lukas Schmitt <schmitt@w3p.cc>
 * Date: 17.04.13
 * Time: 16:04
 *
 */
var
    storage = require("../"),
    rclient = storage.getConnection(),
    Int64 = require('node-int64');

module.exports = {
    testClearAllData: function(test){
        storage.clearAllData(function(){
            rclient.KEYS('app*', function(err, appKeys){
                test.equal(appKeys.length, 0, "Not all keys are deleted");
                test.done();
            });
        });
    },

    testInit1: function (test) {
        var data1 = {
            protocolVersion: 0x01,
            username: "test",
            password: "test",
            app: "Test Applikation",
            processes: 4,
            nodes: 2,
            nodeId: 0,
            ranks: [
                0, 1
            ],
            counters: [
                "broadcast",
                "barrier"
            ],
            sends: [
                "ibsend",
                "bsend",
                "irsend"
            ],
            features: {
                restart: true,
                abort: true
            }
        };

        console.log("Init data1");
        storage.initAction(null, data1, function(err, appId){
            test.notEqual(appId, null, "AppId is undefined");
            test.done();

        });
    },
    testInit2: function(test){
        var data2 = {
            protocolVersion: 0x01,
            username: "test",
            password: "test",
            app: "Test Applikation",
            processes: 4,
            nodes: 2,
            nodeId: 1,
            ranks: [
                2, 3
            ]
        };
        console.log("Init data2");
        storage.initAction(0, data2, function(err, appId){
            test.equal(appId, 0, "AppId is not 0");
            test.done();
        });
    },


};
