// ompimon-storage/test/02_clusterDataTest.js
/**
 * @author Lukas Schmitt <schmitt@w3p.cc>
 * Date: 24.04.13
 * Time: 17:05
 *
 */
var
    storage = require("../"),
    _ = require("underscore")._,
    redis = require("redis"),
    rclient = redis.createClient(),
    Int64 = require('node-int64');

module.exports = {
    testData1: function(test){
        var data = [
            {
                rank: 0,
                counters: {
                    broadcast: 512,
                    barrier: 123
                },
                data: [
                    {
                        rank: 1,
                        counter: new Int64(12344),
                        size: new Int64(237373)
                    },
                    {
                        rank: 2,
                        counter: new Int64(123123),
                        size: new Int64(2020203)
                    },
                    {
                        rank: 3,
                        counter: new Int64(123123),
                        size: new Int64(2020203)
                    }
                ]
            },
            {
                rank: 1,
                counters: {
                    broadcast: 123,
                    barrier: 567
                },
                data: [
                    {
                        rank: 0,
                        counter: new Int64(12944),
                        size: new Int64(207373)
                    },
                    {
                        rank: 2,
                        counter: new Int64(121123),
                        size: new Int64(2020103)
                    },
                    {
                        rank: 3,
                        counter: new Int64(122123),
                        size: new Int64(2020213)
                    }
                ]
            }
        ];
        var appId = 0;
        storage.receiveAction(appId, data, function(){
            var counterPrefix = "app:"+appId+":rank:"+data[0].rank+":counter:";
            rclient.HGET(counterPrefix+"broadcast:value:1", "value",function(err, value){
                test.equal(data[0].counters.broadcast, value, "Wrong value for broadcast");
                test.done();
            });

        });

    },
    testData2: function(test){
        var data = [
            {
                rank: 0,
                counters: {
                    broadcast: 861,
                    barrier: 541
                },
                data: [
                    {
                        rank: 1,
                        counter: new Int64(25217),
                        size: new Int64(394456)
                    },
                    {
                        rank: 2,
                        counter: new Int64(217239),
                        size: new Int64(3526935)
                    },
                    {
                        rank: 3,
                        counter: new Int64(258382),
                        size: new Int64(3129563)
                    }
                ]
            },
            {
                rank: 1,
                counters: {
                    broadcast: 411,
                    barrier: 629
                },
                data: [
                    {
                        rank: 0,
                        counter: new Int64(22944),
                        size: new Int64(307373)
                    },
                    {
                        rank: 2,
                        counter: new Int64(221123),
                        size: new Int64(3020103)
                    },
                    {
                        rank: 3,
                        counter: new Int64(221193),
                        size: new Int64(3021213)
                    }
                ]
            }
        ];
        var appId = 0;
        storage.receiveAction(appId, data, function(){
            test.done();
        });

    }
};
