// ompimon-storage/test/04_getDataTest.js
var
    storage = require("../"),
    rclient = storage.getConnection();

var appId = 0;

module.exports = {
    testLastCounterData: function(test){
        var data1 = 861;
        var data2 = 411;
        var counterId = 0; //broadcast
        storage.getLastCounterData(appId, counterId, function(err, list){
            test.equal(data1, list[0].value, "Wrong value for broadcast rank 0");
            test.equal(data2, list[1].value, "Wrong value for broadcast rank 1");
            test.done();
        });

    },

    testLastCounterData2: function(test){
        var length = 4;
        var counterId = 0; //broadcast
        storage.getLastCounterData(appId, counterId, function(err, list){
            //console.dir(list);
            test.equal(length, list.length, "Wrong size of Array");
            test.done();
        });

    },

    testCounterData: function(test){
        var counterId = 1; //broadcast
        var rankId = 0;
        var data = 123;
        storage.getCounterData(appId, counterId, rankId, function(err, list){
            test.equal(data, list[1].value, "Wrong value for broadcast rank 1");
            test.done();
        });

    },

    testLastSendData: function(test){
        var sendId = null; //broadcast
        var data1 = 394456;
        var data2 = 307373;

        storage.getLastSendData(appId, sendId, function(err, list){
            //console.dir(list);
            test.equal(data1, list[0].size, "Wrong value for send rank 0 to 1");
            test.equal(data2, list[3].size, "Wrong value for send rank 0 to 1");
            test.done();
        });

    },

    testSendData: function(test){
        var data = 207373;
        storage.getSendData(appId,null,1, function(err, list){
            //console.dir(list);
            test.equal(data, list[0].size, "Wrong value for send rank 0 to 1");
            test.done();
        });
    },

    testQuitConnection: function(test){
        storage.killConnection();
        test.done();
    }
};
