// ompimon-storage/test/clusterTest.js
var storage = require("../"),
    redis = require("redis"),
    rclient = redis.createClient(),
    Int64 = require('node-int64');

module.exports = {
    testInit1: function (test) {
        storage.clearData();
        var data1 = {
            protocolVersion: 0x01,
            username: "test",
            password: "test",
            app: "Test Applikation",
            processes: new Int64(4),
            nodes: new Int64(2),
            nodeId: 0,
            ranks: [
                1, 2
            ],
            counters: [
                "broadcast",
                "barrier"
            ],
            sends: [
                "ibsend",
                "bsend",
                "irsend"
            ],
            features: {
                restart: true,
                abort: true
            }
        };

        console.log("Init data1");
        storage.InitAction(data1, function(appId){
            test.notEqual(appId, null, "AppId is undefined");
            test.done();
        });
    },
    testInit2: function(test){
        var data2 = {
            protocolVersion: 0x01,
            username: "test",
            password: "test",
            app: "Test Applikation",
            processes: new Int64(4),
            nodes: new Int64(2),
            nodeId: 1,
            ranks: [
                3, 4
            ]
        };
        console.log("Init data2");
        storage.InitAction(data2, function(appId){
            test.equal(appId, 0, "AppId is not 0");
            test.done();
        });
    }

};
